build.gradle

```diff
@@ -1,93 +1,284 @@
-    into to_path_dir
+    String getArchiveFileName() {
+        String suffix
+        switch (goArch) {
+            case "386": suffix = '-32bit'; break;
+            case "amd64": suffix = ''; break;
+            default: suffix = "-${goArch}"; break;
+        }
+        String archivetype = (goOs == 'linux') ? 'tgz' : 'zip'
+        "${zipFileName}-${this.owOs}${suffix}.${archivetype}"
+    }
 }
 
-pushImage.finalizedBy copyCLIShortcut
+/*
+    Configuration of OpenWhisk Platform behavior based on environment and defaults
+ */
+OpenWhiskPlatform.zipFileName =
+    System.env['zip_file_name'] ?:
+        (rootProject.findProperty('zipFileName') ?: 'OpenWhisk_CLI')
 
-// Returns the Go CLI docker build args
-def getDockerBuildArgs() {
-    String local_os = mapOsNameToGoName(getOsName())
-    String local_arch = mapOsArchToGoArch(getOsArch())
-    def res = []
+/*
+    'platforms' property will be null for a local compile, or a list (comma or
+    space-separated) of hyphenated Goos-Goarch pairs.  Some transformation is
+    done when parsing to handle misconceptions.
 
-    if(!project.hasProperty('crossCompileCLI') || project.crossCompileCLI == "false") {
-        res = ["CLI_OS=${local_os}", "CLI_ARCH=${local_arch}"]
+    TODO:  More syntax/validity checking and feedback, perhaps as part of a
+    Platform object as proposed above...
+*/
+rootProject.ext.localPlatform = new OpenWhiskPlatform()
+
+if (rootProject.hasProperty('buildPlatforms')) {
+    rootProject.ext.platforms = buildPlatforms.tokenize(' ,').collect {
+        new OpenWhiskPlatform(it)
+    }
+} else {
+    if (!rootProject.hasProperty('nativeCompile')) {
+        rootProject.ext.platforms = [
+            'linux-386', 'linux-amd64',
+            'linux-s390x', 'linux-ppc64le', 'linux-arm', 'linux-arm64',
+            'darwin-386', 'darwin-amd64',
+            'windows-386', 'windows-amd64'
+        ].collect { new OpenWhiskPlatform(it) }
     } else {
-        res = ["CLI_OS=mac linux windows", "CLI_ARCH=386 amd64"]
+        rootProject.ext.platforms = [ rootProject.localPlatform ]
     }
+}
+
+/*
+    Checks -- add golint and scancode to the checks run prior to build.
+ */
+task goLint(type: com.github.blindpirate.gogradle.Go) {
+    run "golint ${golang.packagePath}"
+}
+
+goCheck.dependsOn(goLint)
+
+goBuild {
+    targetPlatform = rootProject.platforms*.goPlatform
+    def now = new Date().format("yyyy-MM-dd'T'HH:mm:ss.SSSZ")
 
-    return res
+    go(['build',
+      '-ldflags', "-X main.CLI_BUILD_TIME=$now" as String,
+      '-o', './build/${GOOS}-${GOARCH}/wsk${GOEXE}',
+      golang.packagePath ] as List<String>)
 }
 
-def run(cmd, ignoreError = false) {
-    println("Executing '${cmd.join(" ")}'")
-    def proc = cmd.execute()
-    proc.waitFor()
-    if(!ignoreError && proc.exitValue() != 0) {
-        println("Command '${cmd.join(" ")}' failed with exitCode ${proc.exitValue()}")
+task compile(type: Copy, dependsOn: goBuild) {
+    destinationDir = file('./bin')
```

Comment:
Do we really need the bin folder, or can we just rely on the build folder?

Reply:
We would just need to copy the native binary to `build/wsk` here for use with the tests.

Reply:
I like the idea of using the build folder -- one less mouth to feed.  I'll make the adjustment(s).